Resumable store copy

phughk opened the pull request:
This PR introduces resumable store copy in the sense that failed store file download will be retried. Previously all store files would be streamed to the client in one response and if any issue would occur we would have to redo the entire process. The new strategy is as follows:

CLIENT -> Prepare store request
SERVER -> Grabs checkpoint lock and streams all "atomic-files" (count store, legacy index and additional providers). It then returns a response that includes last committed transaction id, a list of store file names, a list of lucene index descriptors.
CLIENT -> for each store file: send get store file request
SERVER -> stream requested file
CLIENT -> for each descriptor: send get lucene snapshot request
SERVER -> stream lucene snapshot files.

If prepare store request fails then store copy fails. Each get file/snapshot request may be retried as long as the configured termination condition allows.

Each file/snapshot request include the "last committed tx id" given by the prepare store copy respons so we can try to download from any instance in the cluster. In this PR all requests are to leader.

As of yet, nothing has been done to "pull tx" which is called after the store files has been copied. This will be investigated in a different PR.

KERNEL:
NeoStoreFileListing have been modified to provide builders for listing files. IndexService has been modified to expose descriptors and to allow getting specific snapshot for descriptor.

PrepareStoreCopyRequestHandler, GetStoreFileRequestHandler and GetLuceneSnapshotRequestHandler all use NeoStoreFileListing and would be good looking into.

phughk commented:
Team-kernel tag added as it would be good to have confirmation that locks are good and all kernel components are used correctly.

chrisvest commented:
Just letting you know that I've started reviewing this. It may take some time before I've made my way through it.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchUpResponseAdaptor.java:
```diff
@@ -35,43 +36,54 @@
+        fail( signal, response );
+    }
+
+    private <U> void fail( CompletableFuture<T> signal, U response )
```
Maybe call this something else to avoid confusing it with the JUnit `fail` method.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -808,9 +809,36 @@ public InwardKernel getKernel()
     }
 
-    public ResourceIterator<StoreFileMetadata> listStoreFiles( boolean includeLogs ) throws IOException
+    public ResourceIterator<StoreFileMetadata> listAllStoreFiles( boolean includeLogs ) throws IOException
```
I'm not convinced about the value of this rename.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -808,9 +809,36 @@ public InwardKernel getKernel()
+
+    public NeoStoreFileListing.ExclusiveFileLister listStoreFilesExclusively()
+    {
+        return getNeoStoreFileListing().exclusiveBuilder();
```
Do we really need two builders?

chrisvest replied:
You could add `includeAll` / `excludeAll` methods to the builder.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -808,9 +809,36 @@ public InwardKernel getKernel()
+        return kernelModule.fileListing();
+    }
+
+    public NeoStoreFileIndexListing getNeoStoreIndexListing()
```
Why are the indexes separate in this way?

RagnarW replied:
As in, why not only have NeoStoreFileListing getter and then get it from there?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+        {
+            File storeDir = dataSource.get().getStoreDir();
+            ResourceIterator<StoreFileMetadata> resourceIterator =
+                    dataSource.get().getNeoStoreIndexListing().getSnapshot( snapshotRequest.descriptor() );
```
Contrary to the name of this class, this will also get native indexes.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -634,16 +637,36 @@ private void closeAllIndexes()
     }
 
-    public ResourceIterator<File> snapshotStoreFiles() throws IOException
+    public Set<Descriptor> getIndexDescriptors()
```
These are not index descriptors, but schema index provider descriptors. Why do we need the schema index provider descriptors? Do you want to resume per provider?

chrisvest replied:
Also, we never need the information that this is a set. It's turned into a Collection by the caller.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -634,16 +637,36 @@ private void closeAllIndexes()
+        return fromProviders;
+    }
+
+    public ResourceIterator<File> snapshotStoreFiles( Predicate<Descriptor> filter ) throws IOException
```
I'd say that these are index files rather than store files.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/MultiResource.java:
```diff
@@ -0,0 +1,55 @@
+        if ( exception != null )
+        {
+            throw exception;
+        }
```
`IOUtils.close`

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/MultiResource.java:
```diff
@@ -0,0 +1,55 @@
+
+public final class MultiResource implements Resource
+{
+    private final Collection<? extends Resource> snapshots;
```
These are not snapshots, but resources.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/MultiResource.java:
```diff
@@ -0,0 +1,55 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.transaction.state;
```
I think `org.neo4j.kernel.impl.util` is a better home. This thing is pretty generic.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/NeoStoreFileListing.java:
```diff
@@ -144,6 +142,185 @@ private void gatherNonRecordStores( Collection<StoreFileMetadata> files, boolean
+        ExclusiveFileLister excludeExplicitIndexStoreFiles();
+        ExclusiveFileLister excludeAdditionalProviders();
+        ResourceIterator<StoreFileMetadata> build() throws IOException;
+    }
```
Can't we get away with one builder, and `excludeAll` / `includeAll` methods? I also think there's no need for the interfaces, when you can expose a `final` class in this case.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/NeoStoreFileListing.java:
```diff
@@ -144,6 +142,185 @@ private void gatherNonRecordStores( Collection<StoreFileMetadata> files, boolean
+
+            placeMetaDataStoreLast( files );
+
+            return resourceIterator( files.iterator(), new MultiResource( resources ) );
```
You have two identical `MultiResource` classes. The one embedded in this class, and the one you copied for the indexes. Please remove one.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -634,16 +637,36 @@ private void closeAllIndexes()
+            Descriptor providerDescriptor = indexProxy.getProviderDescriptor();
+            if ( filter.test( providerDescriptor ) )
+            {
+                snapshots.add(  indexProxy.snapshotFiles() );
```
Formatting.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/NeoStoreFileIndexListing.java:
```diff
@@ -0,0 +1,104 @@
+    private final IndexingService indexingService;
+    private final ExplicitIndexProviderLookup explicitIndexProviders;
+
+    private final Function<File,StoreFileMetadata> toNotAStoreTypeFile = file -> new StoreFileMetadata( file, RecordFormat.NO_RECORD_SIZE );
```
This can be `static`. I think the `NeoStoreFileIndexingListing` gives enough context that we know these are not store files. The field could instead say what it does rather than what it does not, e.g. `toStoreFileMetadata`.

chrisvest replied:
Isn't this line too long for our code style?

RagnarW replied:
Quite recently I got the news that we use 160 characters in our code style. I had been using 120

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/NeoStoreFileIndexListing.java:
```diff
@@ -0,0 +1,104 @@
+
+    private List<StoreFileMetadata> getSnapshotFilesMetadata( ResourceIterator<File> snapshot )
+    {
+        return snapshot.stream().map( toNotAStoreTypeFile ).collect( Collectors.toList() );
```
You could also take the target `Collection<StoreFileMetadata>` as a parameter and use `forEach( target::add )` to avoid copying via the `toList` collector.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/CloseablesListener.java:
```diff
@@ -0,0 +1,77 @@
+{
+    private final List<AutoCloseable> closeables = new ArrayList<>();
+
+    <T extends AutoCloseable> T sample( T closeable )
```
Why is this called `sample` and not `add`?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/CloseablesListener.java:
```diff
@@ -0,0 +1,77 @@
+        if ( exception != null )
+        {
+            throw exception;
+        }
```
`IOUtils.close`

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+            }
+            catch ( Throwable e )
+            {
+                storeFileStreamingProtocol.end( ctx, StoreCopyFinishedResponse.Status.E_UNKNOWN ).addListener( closeablesListener );
```
You are dropping the `Throwable` on the floor, here. How about keeping the `StoreCopyFinishedResponse.Status` in a local variable, initialised to `E_UNKNOWN`, and then move `storeFileStreamingProtocol.end(…)` to the `finally` clause?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+
+import static org.neo4j.io.fs.FileUtils.relativePath;
+
+public class GetLuceneSnapshotRequestHandler extends SimpleChannelInboundHandler<GetLuceneSnapshotRequest>
```
This should be `GetIndexSnapshotRequestHandler`, and likewise for the `*Request` etc.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+    @Override
+    protected void channelRead0( ChannelHandlerContext ctx, GetLuceneSnapshotRequest snapshotRequest ) throws Exception
+    {
+        log.debug( "Requesting Lucene snapshot %s", snapshotRequest );
```
_index_ snapshot.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+                    StoreFileMetadata storeFileMetadata = resourceIterator.next();
+                    storeFileStreamingProtocol.stream( ctx,
+                            new StoreResource( storeFileMetadata.file(), relativePath( storeDir, storeFileMetadata.file() ), storeFileMetadata.recordSize(),
+                                    pageCache, fs ) );
```
I think you could extract some local variables to shorten this line.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupAddressProvider.java:
```diff
@@ -0,0 +1,114 @@
+{
+    AdvertisedSocketAddress primary() throws CatchupAddressResolutionException;
+
+    AdvertisedSocketAddress secondary() throws CatchupAddressResolutionException;
```
Some javadocs on the interface and methods would be helpful, I think.

chrisvest replied:
Meanwhile, I don't find any usages of the `secondary()` method. Can it be removed (making this a functional interface in the process) or is there some functionality that isn't implemented yet? Or is in the commercial repo?

RagnarW replied:
Yes. Currently we do all request to the leader. We plan to expand on this so that we only the the prepare request to the leader and then the interface will allow us to route the file/snapshot requests to other instances.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupClientProtocol.java:
```diff
@@ -35,6 +35,8 @@ public CatchupClientProtocol()
         TX_STREAM_FINISHED,
         FILE_HEADER,
+        PREPARE_STORE_COPY_RESPONSE,
+        LUCENE_SNAPSHOT_RESPONSE,
```
`s/LUCENE/INDEX/`

Is adding/changing these enums a protocol change? Does it involve some versioning?

RagnarW replied:
It is. AFAIK we will resolve this together with the rolling upgrades feature that Martin and Andrew are working on before 3.4 is released.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupServerProtocol.java:
```diff
@@ -28,6 +28,12 @@ public CatchupServerProtocol()
+        GET_STORE_ID,
+        GET_CORE_SNAPSHOT,
+        GET_STORE_FILE,
+        GET_LUCENE_SNAPSHOT,
```
`s/LUCENE/INDEX/`

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/RequestMessageType.java:
```diff
@@ -26,9 +26,11 @@
 {
     TX_PULL_REQUEST( (byte) 1 ),
-    STORE( (byte) 2 ),
+    PREPARE_STORE_COPY( (byte) 2 ),
```
This looks like a protocol change? I don't know if that matters.

RagnarW replied:
It is. AFAIK we will resolve this together with the rolling upgrades feature that Martin and Andrew are working on before 3.4 is released.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/RequestMessageType.java:
```diff
@@ -26,9 +26,11 @@
     CORE_SNAPSHOT( (byte) 3 ),
     STORE_ID( (byte) 4 ),
+    STORE_FILE( (byte) 6 ),
+    LUCENE_SNAPSHOT( (byte) 7 ),
```
`s/LUCENE/INDEX/`

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/ResponseMessageType.java:
```diff
@@ -29,7 +29,9 @@
     TX_STREAM_FINISHED( (byte) 6 ),
-    UNKNOWN( (byte) 200 ),;
+    PREPARE_STORE_COPY_RESPONSE( (byte) 7 ),
+    LUCENE_SNAPSHOT_RESPONSE( (byte) 8 ),
```
`s/LUCENE/INDEX/`

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequest.java:
```diff
@@ -0,0 +1,135 @@
+        private void putString( WritableChannel channel, String version ) throws IOException
+        {
+            channel.putInt( version.length() );
+            channel.put( version.getBytes(), version.length() );
```
Maybe call it `value` instead of `version`. Also, you are letting the platform environment dictate the string encoding, here. I'd pin that down to UTF-8, so both ends know what to expect.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequest.java:
```diff
@@ -0,0 +1,135 @@
+            long requiredTransactionId = channel.getLong();
+            byte[] descriptorKey = getBytes( channel );
+            byte[] descriptorVersion = getBytes( channel );
+            return new GetLuceneSnapshotRequest( storeId, new SchemaIndexProvider.Descriptor( new String( descriptorKey ), new String( descriptorVersion ) ),
```
You are letting the platform environment dictate the encoding in your `new String`s here.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponse.java:
```diff
@@ -0,0 +1,230 @@
+        }
+
+        @Override
+        protected PrepareStoreCopyResponse unmarshal0( ReadableChannel channel ) throws IOException, EndOfStreamException
```
`EndOfStreamException` is never thrown in this method, as far as I can see.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponse.java:
```diff
@@ -0,0 +1,230 @@
+            buffer.putInt( files.length );
+            for ( File file : files )
+            {
+                byte[] bytes = file.getName().getBytes();
```
Give `getBytes` a `Charset`, for instance `StandardCharsets.UTF_8`, to avoid the platform encoding dictating what this produces.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponse.java:
```diff
@@ -0,0 +1,230 @@
+                buffer.putInt( ketBytes.length );
+                buffer.put( ketBytes, ketBytes.length );
+
+                byte[] versionBytes = descriptor.getVersion().getBytes();
```
Same encoding issue for key and version.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetStoreFileRequest.java:
```diff
@@ -0,0 +1,119 @@
+            int fileNameLength = channel.getInt();
+            byte[] fileNameBytes = new byte[fileNameLength];
+            channel.get( fileNameBytes, fileNameLength );
+            return new GetStoreFileRequest( storeId, new File( new String( fileNameBytes ) ), requiredTransactionId );
```
`fileNameBytes` is interpreted according to platform encoding. Give it a `StandardCharsets.UTF_8` to pin it down.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetStoreFileRequest.java:
```diff
@@ -0,0 +1,119 @@
+            channel.putLong( getStoreFileRequest.requiredTransactionId() );
+            String name = getStoreFileRequest.file().getName();
+            channel.putInt( name.length() );
+            channel.put( name.getBytes(), name.length() );
```
Platform encoding again, on `name.getBytes()`.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponse.java:
```diff
@@ -0,0 +1,230 @@
+            int characters = channel.getInt();
+            byte[] name = new byte[characters];
+            channel.get( name, characters );
+            return new File( new String( name ) );
```
Again with the platform encoding.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponse.java:
```diff
@@ -0,0 +1,230 @@
+            int versionCharacters = channel.getInt();
+            byte[] version = new byte[versionCharacters];
+            channel.get( version, versionCharacters );
+            return new SchemaIndexProvider.Descriptor( new String( key ), new String( version ) );
```
Again with the platform encoding.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponse.java:
```diff
@@ -0,0 +1,230 @@
+            return new File( new String( name ) );
+        }
+
+        private SchemaIndexProvider.Descriptor[] unmarshalIndexDescriptors( ReadableChannel channel ) throws IOException
```
I'd call these index provider descriptors.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClient.java:
```diff
@@ -19,63 +19,132 @@
 
-    public long copyStoreFiles( AdvertisedSocketAddress from, StoreId expectedStoreId, StoreFileStreams storeFileStreams ) throws StoreCopyFailedException
+    public long copyStoreFiles( CatchupAddressProvider catchupAddressProvider, StoreId expectedStoreId, StoreFileStreams storeFileStreams,
+            Supplier<TerminationCondition> requestWiseTerminationCondition )
```
The `StoreCopyClient` is always created together with `RemoteStore`, which is the only class that calls this method and passes  a `requestWiseTerminationCondition`. I wonder if this argument could be made a field instead, since this argument list is already quite long. And perhaps `RemoteStore` could be put in charge of creating the `StoreCopyClient`?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyResponseAdaptor.java:
```diff
@@ -0,0 +1,71 @@
+    private final Log log;
+    private String destination;
+    private int requiredAlignment;
+    private Set<String> debugCopiedFiles = new HashSet<>();
```
I don't think we need this field.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/TerminationCondition.java:
```diff
@@ -0,0 +1,34 @@
+
+interface TerminationCondition
+{
+    void assertContinue() throws StoreCopyFailedException;
```
This can be annotated as a functional interface. It would be good to add some javadocs.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/TerminationCondition.java:
```diff
@@ -0,0 +1,34 @@
+{
+    void assertContinue() throws StoreCopyFailedException;
+
+    class ContinueIndefinitely implements TerminationCondition
```
This class is only used in tests. It could be inlined as lambdas there, or you could have a static method in the interface, that returns a do-nothing lambda implementation. That way you won't need to name and expose a concrete type for this.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/NeoStoreFileListing.java:
```diff
@@ -144,6 +142,185 @@ private void gatherNonRecordStores( Collection<StoreFileMetadata> files, boolean
+        }
+    }
+
     private Resource gatherExplicitIndexFiles( Collection<StoreFileMetadata> files ) throws IOException
```
This and the following two private methods are no longer used.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/NeoStoreFileListing.java:
```diff
@@ -144,6 +142,185 @@ private void gatherNonRecordStores( Collection<StoreFileMetadata> files, boolean
+        public ResourceIterator<StoreFileMetadata> build() throws IOException
+        {
+            List<StoreFileMetadata> files = new ArrayList<>();
+            List<Resource> resources = new ArrayList<>();
```
We should make sure that the resources we've collected, are closed if that `IOException` is thrown mid-way through.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/StoreListingResponseHandler.java:
```diff
@@ -0,0 +1,46 @@
+public class StoreListingResponseHandler extends SimpleChannelInboundHandler<PrepareStoreCopyResponse>
+{
+    private final CatchupClientProtocol protocol;
+    private CatchUpResponseHandler handler;
```
Any reason `handler` is not also `final`?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    private boolean isTransactionWithinReach( long atLeast )
```
Perhaps `requiredTxId` is more descriptive of what this is, than `atLeast`.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    private boolean isEqualOrGreater( long atLeast )
```
This method could be called `isWithinLastCheckPoint`, to make it sound more domain relevant.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+            }
+            catch ( IOException e )
+            {
+                log.error( "Unable to acquire lock while preparing to serve catchup, consider the the transaction to be `behind`" );
```
Is this really an error, or can we log it as a warning instead? The fallback is that we'll do a full copy instead of an incremental one, right? Logging with `error` could potentially mean that an ops person gets woken up at night.

RagnarW replied:
Fallback is that client will retry somewhere else if it is allowed.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetStoreFileRequestHandler.java:
```diff
@@ -0,0 +1,149 @@
+                            throw new IllegalStateException( "Found multiple files of: " + getFileName( fileRequest ) );
+                        }
+                        return storeFileMetadata.get( 0 );
+                    } ) );
```
I think it reads better if you do these checks outside of the collector.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetStoreFileRequestHandler.java:
```diff
@@ -0,0 +1,149 @@
+    private boolean isEqualOrGreater( long atLeast )
+    {
+        return checkpointerSupplier.get().lastCheckPointedTransactionId() >= atLeast;
+    }
```
The `isTransactionWithinReach` and `isEqualOrGreater` methods are also found in the `GetLuceneSnapshotRequestHandler` class. Could we find a way to only have one copy?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/GetLuceneSnapshotRequestHandler.java:
```diff
@@ -0,0 +1,134 @@
+
+    private boolean isEqualOrGreater( long atLeast )
+    {
+        return checkpointerSupplier.get().lastCheckPointedTransactionId() >= atLeast;
```
This is true if the most recent checkpoint transaction at the time of the "prepare store copy" is still the most recent one at this time, or if a more recent checkpoint exists. It will be false if the most recent checkpoint is now older than the most recent checkpoint observed at the "prepare store copy" time.

What's the logic behind this check? The `isTransactionWithinReach` sounds like we are asserting that the transaction log is still containing the transaction pointed to by the latest checkpoint at "prepare store copy" time, but I'm not sure that this check accomplishes this.

RagnarW replied:
If this is false the request will fail and the client will have to retry somewhere else. We do not try to achieve that this is true its only a condition for being able to continue.

Currently all requests are to the leader so this should always be true. But we want to be able to send requests to other instances, possible we will have to force checkpointing to happen to avoid this condition constantly failing?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/MaximumTotalRetries.java:
```diff
@@ -0,0 +1,64 @@
+
+    private boolean timeHasExpired( long previousCheck, long currentTime )
+    {
+        return (currentTime - previousCheck) > allowedInBetweenTimeMillis;
```
This is most likely always true on the _first_ call to `assertContinue`. I don't know if that's intentional.

RagnarW replied:
I think this is correct. The in between time is to avoid increment if there is some  temporary unavailability on the client side.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyFiles.java:
```diff
@@ -0,0 +1,102 @@
+        Stream<StoreFileMetadata> explicitIndexStream =
+                neoStoreDataSource.listStoreFilesInclusively().includeExplicitIndexStoreStoreFiles().includeAdditionalProviders().build().stream();
+        StoreFileMetadata[] metaDataStores =
+                closeablesListener.sample( Stream.concat( countStoreStream, explicitIndexStream ) ).toArray( StoreFileMetadata[]::new );
```
I think you should add the resource iterators to the `closeablesListener` one at a time, so if the second `build()` fails, you don't forget to close the first one. If you add the resource iterators directly, you can then later build the streams, and then use them to map to StoreResources, and then return the final array.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/PrepareStoreCopyRequestHandler.java:
```diff
@@ -0,0 +1,119 @@
+        catch ( Throwable e )
+        {
+            channelHandlerContext.writeAndFlush( PrepareStoreCopyResponse.error( PrepareStoreCopyResponse.Status.E_LISTING_STORE ) ).addListener(
+                    closeablesListener );
```
You're not propagating the `Throwable` here.

chrisvest replied:
You could move `channelHandlerContext.writeAndFlush( success ).addListener( closeablesListener );` to the `finally` clause, and instead just assign `success` in the catch clause, or whatever you want to call that variable.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/RemoteStore.java:
```diff
@@ -53,6 +56,7 @@
  */
 public class RemoteStore
 {
+    public static final Supplier<TerminationCondition> DEFAULT_TERMINATION_CONDITIONS = () -> new MaximumTotalRetries( 10, 10 );
```
This can be private.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClient.java:
```diff
@@ -19,63 +19,132 @@
+                    terminationCondition.assertContinue();
                 }
+            }
+            while ( !successful );
```
If we fail because the `lastTransactionId` is too old, then there's no point in retrying, right? Could we have a case to handle that? And I suppose in that case we'd also have to throw away anything we've already copied thus far, right?

RagnarW replied:
Currently that is true. But we plan to download files from multiple instances and some of them might be falling behind. In that case we want to retry the request against another instance.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClient.java:
```diff
@@ -97,4 +166,61 @@ public void onGetStoreIdResponse( CompletableFuture<StoreId> signal, GetStoreIdR
+        log.debug( "Request for individual file resulted in response type: %s", response.status() );
+        if ( responseStatus == StoreCopyFinishedResponse.Status.E_TOO_FAR_BEHIND )
+        {
+            return false;
```
Regarding my comment above, I guess we'd want to throw a `StoreCopyFailedException` here instead.

RagnarW replied:
Same response as above :)

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloader.java:
```diff
@@ -103,32 +99,30 @@ void downloadSnapshot( MemberId source ) throws StoreCopyFailedException
                 if ( catchupResult == E_TRANSACTION_PRUNED )
                 {
-                    log.info( format( "Failed to pull transactions from %s (%s). They may have been pruned away", source, fromAddress ) );
+                    log.info( format( "Failed to pull transactions from %s (%s). They may have been pruned away", addressProvider, primary ) );
```
Will the address provider have a useful toString in this case?

RagnarW replied:
It should!

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/FakeCatchupServer.java:
```diff
@@ -0,0 +1,217 @@
+            protected void channelRead0( ChannelHandlerContext channelHandlerContext, GetStoreFileRequest getStoreFileRequest ) throws Exception
+            {
+//                log.debug( "Received request for file %s", getStoreFileRequest.file().getName() );
+                log.info( "Received request for file %s", getStoreFileRequest.file().getName() );
```
Debugging leftovers?

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/MaximumTotalRetriesTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    public void shouldContinueIfAlloedInBetweenTimeIsMet() throws Exception
```
`s/Alloed/Allowed/`

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/SimpleCatchupClient.java:
```diff
@@ -0,0 +1,124 @@
+        CloseablesListener closeablesListener = new CloseablesListener();
+        closeablesListener.sample( streamToDisk );
+        closeablesListener.sample( clientPageCache );
+        closeablesListener.close();
```
You could also `IOUtils.closeAll` these.

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClientIT.java:
```diff
@@ -0,0 +1,262 @@
+        try ( StoreChannel storeChannel = fileSystemAbstraction.create( file ) )
+        {
+            storeChannel.write( ByteBuffer.wrap( bytes ) );
+            storeChannel.close();
```
The try-with-resources takes care of the closing, right?

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClientIT.java:
```diff
@@ -0,0 +1,262 @@
+        };
+    }
+
+    private interface ThrowableSupplier<T>
```
We also have `org.neo4j.function.ThrowingSupplier`.

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/TerminationConditionMaximumRetriesTest.java:
```diff
@@ -0,0 +1,69 @@
+        for ( int i = 0; i < MAX_RETRIES - 1; i++ )
+        {
+            subject.assertContinue();
+            Thread.sleep( 2 * DURATION_BETWEEN_RETRIES_SHORT );
```
ಠ_ಠ

Can you somehow use `FakeClock` instead?

RagnarW replied:
This looks to be overlapping MaximumTotalRetriesTest which uses a fake clock. I'll just remove it.

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/storecopy/TestPageCache.java:
```diff
@@ -0,0 +1,58 @@
+import org.neo4j.io.pagecache.tracing.cursor.context.EmptyVersionContext;
+import org.neo4j.io.pagecache.tracing.cursor.context.EmptyVersionContextSupplier;
+
+public class TestPageCache
```
Can't you use `PageCacheRule` or `StandalonePageCacheFactory` instead?

chrisvest reviewed (changes requested):
Don't be alarmed by the number of comments. Most of them are small things.

I think, however, that you are grabbing the wrong kind of descriptors. The schema provider descriptors are for the _schema index providers_, where Lucene is one, the native indexes are one, etc. I think what you want to grab are the _index descriptors_, such that you are grabbing a snapshot, so to speak, per index, rather than for all indexes for a given schema index provider. For instance, if the Lucene schema index provider is servicing all your indexing needs in a particular database, then all of your indexes would be stream as a single snapshot, if I'm reading the code correctly.

RagnarW commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/MaximumTotalRetries.java:
```diff
@@ -49,8 +49,8 @@ public void assertContinue() throws StoreCopyFailedException
             tries++;
-            previousCheck = currentTime;
         }
+        previousCheck = currentTime;
```
This was mistakenly moved

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/IndexDescriptorSerializer.java:
```diff
@@ -0,0 +1,55 @@
+
+class IndexDescriptorSerializer
+{
+    private static final byte LABEL_SCHEMA_DESCRIPTOR = 1;
```
Good future-proving 👍

phughk committed:
Resumable copy

This PR introduces resumable store copy in the sense that failed store
file download will be retried. Previously all store files would be
streamed to the client in one response and if any issue would occur we
would have to redo the entire process. The new strategy is as follows:

CLIENT -> Prepare store request
SERVER -> Grabs checkpoint lock and streams all "atomic-files" (count
store, legacy index and additional providers). It then returns a
response that includes last committed transaction id, a list of store
file names, a list of lucene index descriptors.
CLIENT -> for each store file: send get store file request
SERVER -> stream requested file
CLIENT -> for each descriptor: send get lucene snapshot request
SERVER -> stream lucene snapshot files.

If prepare store request fails then store copy fails. Each get
file/snapshot request may be retried as long as the configured
termination condition allows.

Each file/snapshot request include the "last committed tx id" given by
the prepare store copy respons so we can try to download from any
instance in the cluster. In this PR all requests are to leader.

As of yet, nothing has been done to "pull tx" which is called after the
store files has been copied. This will be investigated in a different
PR.

KERNEL:
NeoStoreFileListing have been modified to provide builders for listing
files. IndexService has been modified to expose descriptors and to allow
getting specific snapshot for descriptor.

PrepareStoreCopyRequestHandler, GetStoreFileRequestHandler and
GetLuceneSnapshotRequestHandler all use NeoStoreFileListing and would be
good looking into.

RagnarW committed:
Keep values for message types.

Add tests to avoid changing response or request messge type values

Add tests to avoid changing order of StoreCopyResponse status and PrepareStoreCopyResponse status

RagnarW committed:
Improve catchup message tests

Add comment to maintain order

Hard code values for message type in tests.
